icicle/appUtils/msm/msm.cu

```diff
@@ -992,21 +799,22 @@ namespace msm {
   template <typename S, typename A, typename P>
   cudaError_t MSM(S* scalars, A* points, int msm_size, MSMConfig& config, P* results)
   {
-    int bitsize = (config.bitsize == 0) ? S::NBITS : config.bitsize;
+    const int bitsize = (config.bitsize == 0) ? S::NBITS : config.bitsize;
     cudaStream_t& stream = config.ctx.stream;
-    // TODO: DmytroTym/HadarIngonyama - unify the implementation of the bucket method and the batched bucket method in
-    // one function
-    if (config.batch_size == 1)
-      return CHK_STICKY(bucket_method_msm(
-        bitsize, 16, scalars, points, msm_size, results, config.are_scalars_on_device,
-        config.are_scalars_montgomery_form, config.are_points_on_device, config.are_points_montgomery_form,
-        config.are_results_on_device, config.is_big_triangle, config.large_bucket_factor, config.is_async, stream));
-    else
-      return CHK_STICKY(batched_bucket_method_msm(
-        bitsize, (config.c == 0) ? get_optimal_c<S>(msm_size) : config.c, scalars, points, config.batch_size, msm_size,
-        (config.points_size == 0) ? msm_size : config.points_size, results, config.are_scalars_on_device,
-        config.are_scalars_montgomery_form, config.are_points_on_device, config.are_points_montgomery_form,
-        config.are_results_on_device, config.is_async, stream));
+
+    unsigned c = config.batch_size > 1 ? ((config.c == 0) ? get_optimal_c(msm_size) : config.c) : 16;
```

Comment:
why is 16 the default for a single msm? it should also be the closest power of 2 to the optimal c

Reply:
It was like that for batch=1 and so I did not want to end up with different behavior.

Should I use the closest power of 2 instead? closest from below? I mean does 12 goes to 8 to 16?

Reply:
I don't remember.. use the value that gives better performance:)

Reply:
I think the optimum is log(N) (meaning the number of ECADD is minimized) but implementation-wise power of 2 work better from what I read. In addition memory grows exponential in c since you have 2^c buckets per BM so practically c=16 is the largest power of two that is reasonable.

Having said that, the code says:
    **unsigned get_optimal_c(int bitsize) { return max((unsigned)ceil(log2(bitsize)) - 4, 1U); }**

I don't understand why '-4'? It is quite odd since I then round it back to a power of two.

Anyway I will try to play with it a little and figure out what value are working better, but I think this is an optimization we can dedicate another PR and task for since I think it's a heuristic and not really part of this PR.

Reply:
yes, for now just change 16 to get_optimal_c. (which will then be rounded to the closest power of 2)
the logn-4 is the value that minimizes the amount of ec additions for most sizes, but it's true that with the actual implementation there are other considerations too. Ideally we will have a dictionary.

Reply:
We discussed in the daily that fine-tuning the value of `c` is probably out of the scope of this PR..
Imo it'll more appropriate to tabulate best values of `c` and optimal accumulation method (triangle vs. non-triangle) when we establish proper benchmarking and run out of options to significantly speed up the MSM (potentially altering choices that we hard-coded). I think for now it's ok to leave reasonable defaults and give users an option to alter them via config

Reply:
I will leave it as is for now then, meaning the behavior remains as it is today.